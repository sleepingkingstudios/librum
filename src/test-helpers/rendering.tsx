import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
