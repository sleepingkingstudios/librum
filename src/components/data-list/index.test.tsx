import * as React from 'react';

import { DataList } from './index';
import { render } from '@test-helpers/rendering';

type ExampleObject = {
  name: string,
  description?: string,
  role?: JSX.Element,
};

describe('<DataList />', () => {
  describe('with data: an empty object', () => {
    const data = {};

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataList data={data} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with data: value', () => {
    const data: ExampleObject = {
      name: 'Kevin Flynn',
      description: null,
      role: (<span className="role">User</span>),
    };

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataList data={data} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with data: value and defaultValue: an element', () => {
    const data: ExampleObject = {
      name: 'Kevin Flynn',
      description: null,
    };
    const defaultValue = (
      <em>unknown</em>
    );

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataList data={data} defaultValue={defaultValue} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with data: value and defaultValue: a string', () => {
    const data: ExampleObject = {
      name: 'Kevin Flynn',
      description: null,
    };
    const defaultValue = 'unknown';

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataList data={data} defaultValue={defaultValue} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
