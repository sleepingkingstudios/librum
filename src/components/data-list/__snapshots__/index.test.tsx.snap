// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataList /> with data: an empty object should match the snapshot 1`] = `
<DocumentFragment>
  <dl
    class="data-list"
  />
</DocumentFragment>
`;

exports[`<DataList /> with data: value and defaultValue: a string should match the snapshot 1`] = `
<DocumentFragment>
  <dl
    class="data-list"
  >
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Name
      </dt>
      <dd
        class="data-list-value"
      >
        Kevin Flynn
      </dd>
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Description
      </dt>
      <dd
        class="data-list-value"
      >
        unknown
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`<DataList /> with data: value and defaultValue: an element should match the snapshot 1`] = `
<DocumentFragment>
  <dl
    class="data-list"
  >
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Name
      </dt>
      <dd
        class="data-list-value"
      >
        Kevin Flynn
      </dd>
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Description
      </dt>
      <dd
        class="data-list-value"
      >
        <em>
          unknown
        </em>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`<DataList /> with data: value should match the snapshot 1`] = `
<DocumentFragment>
  <dl
    class="data-list"
  >
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Name
      </dt>
      <dd
        class="data-list-value"
      >
        Kevin Flynn
      </dd>
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Description
      </dt>
      <dd
        class="data-list-value"
      />
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Role
      </dt>
      <dd
        class="data-list-value"
      >
        <span
          class="role"
        >
          User
        </span>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;
